sources/packages/backend/libs/integrations/src/esdc-integration/application-changes-report/application-changes-report.processing.service.ts

```diff
@@ -0,0 +1,122 @@
+
+@Injectable()
+export class ApplicationChangesReportProcessingService {
+  private readonly esdcConfig: ESDCIntegrationConfig;
+  constructor(
+    private readonly applicationService: ApplicationService,
+    private readonly applicationChangesReportIntegrationService: ApplicationChangesReportIntegrationService,
+    private readonly systemUsersService: SystemUsersService,
+    private readonly studentAssessmentService: StudentAssessmentService,
+    config: ConfigService,
+  ) {
+    this.esdcConfig = config.esdcIntegration;
+  }
+
+  /**
+   * Generate application changes report for the applications which has at least one e-Cert sent
+   * and the application study dates have changed after the first e-Cert
+   * or after the last time the application was reported for study dates change
+   * through application changes report. Once generated upload the report to the ESDC directory
+   * in SFTP server.
+   * @param processSummary process summary.
+   * @returns processing result.
+   */
+  async processApplicationChanges(
+    processSummary: ProcessSummary,
+  ): Promise<ApplicationChangesReportProcessingResult> {
+    processSummary.info(
+      "Retrieving all application changes which were not reported already.",
+    );
+    const applicationChanges =
+      await this.applicationService.getDateChangeNotReportedApplications();
+    const applicationChangesCount = applicationChanges.length;
+    processSummary.info(
+      `Found ${applicationChangesCount} application changes.`,
+    );
+    // Build application changes report CSV file content.
+    const fileContent =
+      this.applicationChangesReportIntegrationService.createApplicationChangesReportFileContent(
+        applicationChanges,
+      );
+    // Create file name and build file path.
+    const { fileName, remoteFilePath } = this.createRequestFileName();
+    // Upload the file to SFTP location.
+    try {
+      await this.applicationChangesReportIntegrationService.uploadRawContent(
+        fileContent,
+        remoteFilePath,
+      );
+    } catch (error: unknown) {
+      // Translate to friendly error message.
+      const errorDescription = `Unexpected error uploading the file: ${remoteFilePath} to SFTP.`;
+      processSummary.error(errorDescription, error);
+      throw new Error(errorDescription, { cause: error });
+    }
+    processSummary.info(
+      `Application changes report with file name: ${fileName} has been uploaded successfully.`,
+    );
+    if (applicationChangesCount) {
+      const reportedAssessmentIds = applicationChanges.map(
+        (applicationChange) => applicationChange.currentAssessment.id,
+      );
+      await this.studentAssessmentService.updateReportedDate(
+        reportedAssessmentIds,
+        new Date(),
+        this.systemUsersService.systemUser.id,
+      );
+      processSummary.info(
+        "Reported date has been successfully updated for reported application assessments.",
```

Comment:
The log message can include the `new Date()` from line 79 to allow an easy association with all records updated.
If we need to get from the DB all records updated as part of this operation we can search by the exact date.

Reply:
Maybe even the modied_at audit column can be set with the same date. - I initially thought about this, but then looking independently from the method perspective thought `reported_date` could be different from updated_at. i.e. considering that method is not aware that `reported_date` is going to be current date and time.